File: doc/api-examples/rust/set-cpuweight.rs
// Copyright Contributors to the Eclipse BlueChi project
//
// SPDX-License-Identifier: MIT-0

use clap::Parser;
use dbus::arg::Variant;
use dbus::blocking::Connection;
use dbus::Path;
use std::time::Duration;

#[derive(Parser)]
struct Cli {
    /// The node name to list the units for
    #[clap(short, long)]
    node_name: String,

    /// The unit name to set the cpu weight for
    #[clap(short, long)]
    unit_name: String,

    /// The new value of the cpu weight
    #[clap(short, long)]
    cpu_weight: u64,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Cli::parse();

    let conn = Connection::new_system()?;

    let bluechi = conn.with_proxy(
        "org.eclipse.bluechi",
        "/org/eclipse/bluechi",
        Duration::from_millis(5000),
    );

    let (node,): (Path,) =
        bluechi.method_call("org.eclipse.bluechi.Controller", "GetNode", (args.node_name,))?;

    let node_proxy = conn.with_proxy("org.eclipse.bluechi", node, Duration::from_millis(5000));

    let new_cpu_weight = Variant(args.cpu_weight);
    let mut values: Vec<(String, Variant<u64>)> = Vec::new();
    values.push(("CPUWeight".to_string(), new_cpu_weight));

    node_proxy.method_call(
        "org.eclipse.bluechi.Node",
        "SetUnitProperties",
        (args.unit_name, false, values),
    )?;

    Ok(())
}

File: doc/api-examples/cpp/set-cpuweight.cpp
/*
 * Copyright Contributors to the Eclipse BlueChi project
 *
 * SPDX-License-Identifier: MIT-0
 */
#include <iostream>
#include <sdbus-c++/sdbus-c++.h>

int main(int argc, char *argv[]) {
        if (argc != 4) {
                std::cerr << "Usage: " << argv[0] << " node_name unit_name cpu_value" << std::endl;
                return 1;
        }

        char *nodeName = argv[1];
        char *unitName = argv[2];
        char *cpuWeight = argv[3];

        auto connection = sdbus::createSystemBusConnection();
        auto controllerProxy = sdbus::createProxy(
                        *connection,
                        sdbus::ServiceName("org.eclipse.bluechi"),
                        sdbus::ObjectPath("/org/eclipse/bluechi"));

        sdbus::ObjectPath nodePath;
        controllerProxy->callMethod("GetNode")
                        .onInterface("org.eclipse.bluechi.Controller")
                        .withArguments(nodeName)
                        .storeResultsTo(nodePath);

        auto nodeProxy = sdbus::createProxy(
                        *connection, sdbus::ServiceName("org.eclipse.bluechi"), sdbus::ObjectPath(nodePath));

        std::vector< sdbus::Struct< std::string, sdbus::Variant > > values;
        u_int64_t val = std::stoi(cpuWeight);
        auto newValue = sdbus::Struct< std::string, sdbus::Variant >("CPUWeight", sdbus::Variant{ val });
        values.push_back(newValue);

        nodeProxy->callMethod("SetUnitProperties")
                        .onInterface("org.eclipse.bluechi.Node")
                        .withArguments(unitName, false, values);

        return 0;
}
